src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java

```diff
@@ -0,0 +1,522 @@
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    FailedNodeException[] failures = new FailedNodeException[nodesIds.length];
+                    for (int i = 0; i < failures.length; i++) {
+                        failures[i] = new FailedNodeException(nodesIds[i], "total failure in fetching store", e);
+                    }
+                    processAsyncFetch(shardId, cacheGeneration, null, failures);
+                }
+            });
+        }
+    }
+
+    /**
+     * The result of a fetch operation. Make sure to first check {@link #hasData()} before
+     * fetching the actual data.
+     */
+    public static class FetchResult<T extends NodeOperationResponse> {
+
+        private final ShardId shardId;
+        private final Map<DiscoveryNode, T> data;
+        private final Set<String> failedNodes;
+
+        public FetchResult(ShardId shardId, Map<DiscoveryNode, T> data, Set<String> failedNodes) {
+            this.shardId = shardId;
+            this.data = data;
+            this.failedNodes = failedNodes;
+        }
+
+        /**
+         * Does the result actually contain data? If not, then there are on going fetch
+         * operations happening, and it should wait for it.
+         */
+        public boolean hasData() {
+            return data != null;
+        }
+
+        /**
+         * Returns the actual data, note, make sure to check {@link #hasData()} first and
+         * only use this when there is an actual data.
+         */
+        public Map<DiscoveryNode, T> getData() {
+            assert data != null : "getData should only be called if there is data to be fetched, please check hasData first";
+            return this.data;
+        }
+
+        /**
+         * Process any changes needed to the allocation based on this fetch result.
+         */
+        public void processAllocation(RoutingAllocation allocation) {
+            for (String failedNode : failedNodes) {
+                allocation.addIgnoreShardForNode(shardId, failedNode);
+            }
+        }
+    }
+
+    /**
+     * A node entry, holding the state of the fetched data for a specific shard
+     * for a giving node.
+     */
+    static class NodeEntry<T> {
+        private final String nodeId;
+        private boolean fetching;
+        @Nullable
+        private T value;
+        private boolean valueSet;
+        private Throwable failure;
+
+        public NodeEntry(String nodeId) {
+            this.nodeId = nodeId;
+        }
+
+        public String getNodeId() {
+            return this.nodeId;
+        }
+
+        public boolean isFetching() {
+            return fetching;
+        }
+
+        public void markAsFetching() {
+            assert fetching == false : "double marking a node as fetching";
+            fetching = true;
+        }
+
+        public void doneFetching(T value) {
```

Comment:
since tis is really an internal class do we really provide public access to all these mutators?